### PR TITLE
MGMT-9763: add resource pkg API to support preflight validations

### PR DESCRIPTION
**Pull request by yevgeny-shnaidman**
The changes in this PR:
1) add API to get a list of Objects from helm YAMLs
2) rename Creator interface to ResourceAPI, and update the code accordingly

### CODE REVIEW

File: pkg/resource/resource.go [hunk 1]
```diff
@@ -159,34 +160,58 @@ func (c *creator) CreateFromYAML(
...
 }
 
+func (r *resource) GetObjectsFromYAML(ctx context.Context, yamlFile []byte) (*unstructured.UnstructuredList, error) {
+	scanner := yamlutil.NewYAMLScanner(yamlFile)
+	objList := &unstructured.UnstructuredList{
+		Items: []unstructured.Unstructured{},
+	}
+	for scanner.Scan() {
+		yamlSpec := scanner.Bytes()
+		obj := &unstructured.Unstructured{
```
Comment (pmtk) [inline pkg/resource/resource.go:170]: Could creating this object on stack (without `&`) have some benefits in relieving GC? Not that this matters much with amount of objects ;)

File: pkg/resource/resource.go [hunk 2]
```diff
@@ -159,34 +160,58 @@ func (c *creator) CreateFromYAML(
...
+			Object: map[string]interface{}{},
+		}
+
+		jsonSpec, err := yaml.YAMLToJSON(yamlSpec)
+		if err != nil {
+			return nil, fmt.Errorf("Could not convert yaml file to json: %s: error %w", string(yamlSpec), err)
```
Comment (pacevedom) [inline pkg/resource/resource.go:176]: Dumping the whole yaml might be hard to read as part of the error message. WDYT?

File: pkg/resource/resource.go [hunk 3]
```diff
@@ -159,34 +160,58 @@ func (c *creator) CreateFromYAML(
...
+		}
+
+		if err = obj.UnmarshalJSON(jsonSpec); err != nil {
+			return nil, fmt.Errorf("cannot unmarshall json spec, check your manifest: %s: %w", jsonSpec, err)
```
Comment (pacevedom) [inline pkg/resource/resource.go:180]: Same here but with json.

File: pkg/resource/resource.go (hunk 2 above)
Comment (yevgeny-shnaidman) [reply pkg/resource/resource.go:176]: We almost never see those errors (yaml mrshall/unmarshal, to json), so in that rare occurrence that it does happen, it worth to have all that data

File: pkg/resource/resource.go (hunk 1 above)
Comment (yevgeny-shnaidman) [reply pkg/resource/resource.go:170]: it does not matter, but it is a good catch, does not make sense to create it as a pointer, and then to derefernce it

Comment (yevgeny-shnaidman) [reply pkg/resource/resource.go]: fixed

File: pkg/resource/resource.go (hunk 2 above)
Comment (qbarrand) [reply pkg/resource/resource.go:176]:
```suggestion
			return nil, fmt.Errorf("could not convert YAML file to JSON %s: error %w", string(yamlSpec), err)
```

File: pkg/resource/resource.go (hunk 3 above)
Comment (qbarrand) [reply pkg/resource/resource.go:180]:
```suggestion
			return nil, fmt.Errorf("cannot unmarshal JSON spec, check your manifest: %s: %w", jsonSpec, err)
```

File: pkg/resource/resource.go [hunk 4]
```diff
@@ -159,34 +160,58 @@ func (c *creator) CreateFromYAML(
 	return nil
 }
 
+func (r *resource) GetObjectsFromYAML(ctx context.Context, yamlFile []byte) (*unstructured.UnstructuredList, error) {
```
Comment (qbarrand) [inline pkg/resource/resource.go:163]: I don't see `ctx` being used in the function.
```suggestion
func (r *resource) GetObjectsFromYAML(yamlFile []byte) (*unstructured.UnstructuredList, error) {
```

File: pkg/resource/resource_test.go [hunk 5]
```diff
@@ -273,7 +273,75 @@ spec:
...
+`)
+	It("good flow", func() {
+		r := NewResourceAPI(kubeClient, metricsClient, pollActions, kernelData, scheme, mockLifecycle, proxyAPI, helper)
+		resultList, err := r.GetObjectsFromYAML(context.TODO(), yamlSpec)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(len(resultList.Items)).To(Equal(2))
+		for _, item := range resultList.Items {
+			Expect(item.GetName()).To(BeElementOf([]string{"dsName", "podName"}))
+			Expect(item.GetKind()).To(BeElementOf([]string{"DaemonSet", "Pod"}))
+		}
```
Comment (qbarrand) [inline pkg/resource/resource_test.go:340]: I think the order should be deterministic here?
```suggestion
		Expect(resultList.Items[0].GetName()).To(Equal("podName"))
		Expect(resultList.Items[0].GetKind()).To(Equal("Pod"))
		Expect(resultList.Items[1].GetName()).To(Equal("dsName"))
		Expect(resultList.Items[1].GetKind()).To(Equal("DaemonSet"))
```

**Review by qbarrand — COMMENTED** [review]
LGTM, just the usual pickiness.

File: pkg/resource/resource.go (hunk 2 above)
Comment (yevgeny-shnaidman) [reply pkg/resource/resource.go:176]: fixed

File: pkg/resource/resource.go (hunk 3 above)
Comment (yevgeny-shnaidman) [reply pkg/resource/resource.go:180]: fixed

File: pkg/resource/resource.go (hunk 4 above)
Comment (yevgeny-shnaidman) [reply pkg/resource/resource.go:163]: fixed

File: pkg/resource/resource_test.go (hunk 5 above)
Comment (yevgeny-shnaidman) [reply pkg/resource/resource_test.go]: fixed

**Pull request merged**